### PR TITLE
(WIP) Implement getVariations method

### PR DESCRIPTION
**Pull request by brandonhudavid**
## Test plan
* Unit tests
* Integration tests

## Issues
- One of the [SDK convenience APIs](https://docs.google.com/document/d/1MCsMioa4TNImryaDdOJJsZwL9j5zFvFIXODArHF3ORs/edit#)

### CODE REVIEW

**Comment by aliabbasrizvi** [comment]
Thanks for sending this in. Will review once outstanding comments are addressed.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 1]
```diff
@@ -411,6 +411,44 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+        return null;
+      }
+
+      var configObj = this.projectConfigManager.getConfig();
+      if (!configObj) {
+        return null;
+      }
+
+      var variationKeys = {};
+      var experiments = Object.values(configObj.experimentKeyMap);
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Add a method to the project config class to get a list of experiments instead of referencing its map directly

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 2]
```diff
@@ -411,6 +412,53 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+    // Ensure that experimentKeys is an array.
+    experimentKeysValidator.validate(experimentKeys);
+    // if experimentKeys is an empty array, return the bucketed variations for all experiments.
+    if (experimentKeys.length === 0) {
+      experimentKeys = getExperimentKeys(projectConfig);
+    }
+    var experimentsToVariations = {};
+    for (i = 0; i < experimentKeys.length; i++) {
+      var experimentKey = experimentKeys[i];
+      var variationKey = this.decisionService.getVariation(configObj, experimentKey, userId, attributes);
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: What happens if the `experimentKey` is invalid, as in, it doesn't exist in the datafile? Can you add unit tests for that scenario as well

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 3]
```diff
@@ -411,6 +412,53 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+/**
+ * Returns an object mapping experiment keys to the respective
+ * variation that the user was bucketed into. If an empty array
+ * is provided for experimentKeys, return the bucketed variations
+ * for all experiment keys. If a bucketing ID is provided in the
+ * attributes parameter, all experiments will use the bucketing ID.
+ * @param  {string}      userId
+ * @param  {Object}      attributes
+ * @param  {Array}       experimentKeys
+ * @return {Object}      experimentsToVariations
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Instead of `experimentsToVariations` just describe it as `Map of experiment key to variation key`

File: packages/optimizely-sdk/lib/core/project_config/index.js [hunk 4]
```diff
@@ -130,6 +130,20 @@ module.exports = {
     return experiment.id;
   },
 
+  /**
+   * TODO: Add doctest
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/core/project_config/index.js]: Yes, add docs 😄

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 5]
```diff
@@ -3470,8 +3588,24 @@ describe('lib/optimizely', function() {
...
+      };
+      assert.deepStrictEqual(variations, optlyInstance.getAllVariations(
+        'testUser',
+        userAttributesWithBucketingId
+      ));
+    });
+
     it('confirm that invalid experiment with the bucketing ID returns null', function() {
-      assert.strictEqual(null, optlyInstance.getVariation(
+      assert.equal(null, optlyInstance.getVariation(
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: Why is this change needed?

File: packages/optimizely-sdk/lib/utils/experiment_keys_validator/index.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright 2016, 2018-2019, Optimizely
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/utils/experiment_keys_validator/index.js]: Should only say 2019

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 7]
```diff
@@ -411,6 +411,44 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
 };
 
+Optimizely.prototype.getAllVariations = function(userId, attributes) {
+  try {
+    if (!this.__isValidInstance()) {
+      this.logger.log(LOG_LEVEL.ERROR, sprintf(LOG_MESSAGES.INVALID_OBJECT, MODULE_NAME, 'getAllVariations'));
+      return null;
+    }
+
+    try {
```
[the remark this answers is not in the document]
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Removed nested try-catch. Resolved

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 1 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Added method getExperimentKeys. Resolved

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 2 above)
Comment (YuanC) [reply packages/optimizely-sdk/lib/optimizely/index.js]: I'll be sure to add that as a scenario

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 8]
```diff
@@ -411,6 +411,44 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
   }
 };
 
+Optimizely.prototype.getAllVariations = function(userId, attributes) {
+  try {
+    if (!this.__isValidInstance()) {
+      this.logger.log(LOG_LEVEL.ERROR, sprintf(LOG_MESSAGES.INVALID_OBJECT, MODULE_NAME, 'getAllVariations'));
+      return null;
```
[the remark this answers is not in the document]
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Refactored in `getVariations` to return an empty object. Resolved

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 9]
```diff
@@ -411,6 +411,44 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
   }
 };
 
+Optimizely.prototype.getAllVariations = function(userId, attributes) {
```
[the remark this answers is not in the document]
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Added docstring for `getVariations`, includes an `experimentKeys` parameter. If `experimentKeys` is an empty array, returns bucketed variations for all experiment keys.

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 2 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Implemented `getVariations` such that, if `experimentKey` is invalid, the `experimentKey` maps to `null` in the returned `experimentsToVariations` map. We also create and log an `INVALID_EXPERIMENT_KEY` error message.

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 3 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Replaced `experimentsToVariations` with `Map of experiment key to variation key`. Resolved

File: packages/optimizely-sdk/lib/utils/experiment_keys_validator/index.js (hunk 6 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/utils/experiment_keys_validator/index.js]: Changed to `Copyright 2019, Optimizely`. Resolved

File: packages/optimizely-sdk/lib/core/project_config/index.js [hunk 10]
```diff
@@ -130,6 +130,22 @@ module.exports = {
...
   },
 
+  /**
+   * Get all experiment keys in the project config
+   * @param  {Object} projectConfig Object representing project configuration
+   * @return {Array}  Array of all experiment keys
+   */
+  getExperimentKeys: function(projectConfig) {
+    var experimentKeys = [];
+    var experiments = Object.values(projectConfig.experimentKeyMap);
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/core/project_config/index.js]: This module [supports any ES5-compliant environment](https://github.com/optimizely/javascript-sdk/tree/master/packages/optimizely-sdk#prerequisites). We can't use `Object.values` because it's part of a newer-than-ES5 version of JavaScript. Another option is to use `Object.keys`, and access each property value using the corresponding property name.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 11]
```diff
@@ -411,6 +412,64 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+ * variation that the user was bucketed into. If an empty array
+ * is provided for experimentKeys, returns the bucketed variations
+ * for all experiment keys. If a bucketing ID is provided in the
+ * attributes parameter, all experiments will use the bucketing ID.
+ * @param  {string}      userId
+ * @param  {Object}      attributes
+ * @param  {Array}       experimentKeys
+ * @return {Object}      Map of experiment key to variation key
+ */
+Optimizely.prototype.getVariations = function(userId, attributes, experimentKeys) {
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/optimizely/index.js]: - The PR title says `getAllVariations`, but this is `getVariations` - which one is correct?
- I prefer having arguments in this order: `experimentKeys, userId, attributes`. This is more consistent with other methods like `activate`, and `getVariation`.

File: packages/optimizely-sdk/lib/optimizely/index.tests.js (hunk 5 above)
Comment (YuanC) [reply packages/optimizely-sdk/lib/optimizely/index.tests.js]: Reverted!

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 12]
```diff
@@ -411,6 +412,65 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+    // if experimentKeys is an empty array, return the bucketed variations for all experiments.
+    var allExperimentKeys = projectConfig.getExperimentKeys(configObj);
+    if (experimentKeys.length === 0) {
+      experimentKeys = allExperimentKeys;
+    }
+    var experimentsToVariations = {};
+    for (i = 0; i < experimentKeys.length; i++) {
+      var experimentKey = experimentKeys[i];
+      // check if experimentKey is in project config
+      if (!allExperimentKeys.includes(experimentKey)) {
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Same issue here with `Array.prototype.includes` - it's not part of ES5, thus we can't use it.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 13]
```diff
@@ -411,6 +412,65 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+        experimentsToVariations[experimentKey] = '';
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Is this supposed to be assigning a value of `null` instead of `''`?

File: packages/optimizely-sdk/lib/core/project_config/index.js (hunk 4 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/core/project_config/index.js]: Added docstring for `getExperimentKeys`; resolved

File: packages/optimizely-sdk/package-lock.json [hunk 14]
```diff
@@ -264,7 +264,7 @@
       "dependencies": {
         "acorn": {
           "version": "3.3.0",
-          "resolved": "https://registry.npmjs.org/acorn/-/acorn-3.3.0.tgz",
```
Comment (mjc1283) [inline packages/optimizely-sdk/package-lock.json]: What's the goal of making these changes in package-lock.json?

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 13 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: I originally assigned a value of `null`, but when checking in the unit tests the value would always be `[null]`. Do you know the reason for `null` being put in an array? If the array could be removed, I'll definitely assign the value to `null`

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 11 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: - getVariations is correct; when we added the experimentKeys parameter, it no longer made sense to call the method getAllVariations
- Would having the parameters in that order be an issue for other languages where it would be more idiomatic for experimentKeys to be an optional parameter?

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 13 above)
Comment (brandonhudavid) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Never mind, unit tests pass when I assign value to `null`. Will resolve in next commit

**Comment by brandonhudavid** [comment]
Closed due to `getVariations` and other SDK convenience APIs being put on hold a la SDK CoP.
